File: inst/midicallback.m
## Copyright (C) 2022 John Donoghue <john.donoghue@ieee.org>
## 
## This program is free software: you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
## 
## This program is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
## 
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see
## <https://www.gnu.org/licenses/>.

## -*- texinfo -*- 
## @deftypefn {} {@var{oldhandle} =} midicallback (@var{midicontrolsObj}, @var{functionHandle})
## @deftypefnx {} {@var{oldhandle} =} midicallback (@var{midicontrolsObj}, [])
## @deftypefnx {} {@var{currhandle} =} midicallback (@var{midicontrolsObj})
## Get, set or clear the midicontrol object callback.
## 
## @subsubheading Inputs
## @var{midicontrolObj} - control object created using midicontrols.
##
## @var{functionHandle} - function handle to set for call back. If it is [],
## the callback function will be cleared.
##
## @strong{NOTE}: currently anonymous functions will not work.
##
## @strong{NOTE}: callbacks should be cleared before losing all references to the midicontrols 
## object.
##
## @subsubheading Outputs
## @var{oldhandle} The previously set midicallback function handle when setting a new callback.
##
## @var{currhandle} The current set midicallback function handle.
##
## @subsubheading Examples
## Set a callback on a midicontrols object
## @example
## ctrl = midicontrols(2001)
## function dispCallback(ctrl),disp(midiread(ctrl)),endfunction;
## midicallback(ctrl, @@dispCallback);
## @end example
##
## Clear the callback on a midicontrols object
## @example
## ctrl = midicontrols(2001)
## midicallback(ctrl, []);
## @end example
##
## Get the current callback on a midicontrols object
## @example
## ctrl = midicontrols(2001)
## cb = midicallback(ctrl);
## @end example
##
## @seealso{midicontrols, midisync, midiread}
## @end deftypefn
function ret = midicallback (ctrlobj, varargin)
  if nargin < 1 || !isa (ctrlobj, "midicontrols")
    error ("Expected midicontrols object");
  endif

  if nargin > 1
    cb = varargin{1};
    if isempty(cb)
       callback = "";
    elseif strcmp(class(cb), "function_handle")
       info = functions(cb);
       if strcmp(info.type, "anonymous")
         error ("anonymous are currently not supported");
       endif
       callback = func2str(cb);
    elseif ischar (cb)
       callback = cb;
    else
       error ("Callback should be a string or function handle or []")
    endif

    ret =   ctrlobj.callback(callback);
  else
    ret =   ctrlobj.callback();
  endif
endfunction

%!function testcallback(ctrl)
%! # do nothing
%!endfunction

%!test
%! a = midicontrols();
%! assert(isa(a, "midicontrols"));
%! assert(midicallback(a), []);
%! assert(midicallback(a, @testcallback), []);
%! assert(midicallback(a), func2str(@testcallback));
%! midicallback(a, []);
%! assert(midicallback(a), []);

%!error <Expected midicontrols object> midicallback(0);

File: src/midicallback.cc
// Copyright (C) 2022 John Donoghue <john.donoghue@ieee.org>
//
// This program is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation; either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program; if not, see <http://www.gnu.org/licenses/>.

// Octave Includes
#include <octave/oct.h>
#include <octave/defun-dld.h>
#include <octave/ov.h>

#include <iostream>
#include <iomanip>

#if defined (HAVE_CONFIG_H)
#  include "config.h"
#endif

#include "midi_object.h"
#include "midi.h"

// PKG_ADD: autoload ("__midicallback__", "__midi__.oct");
DEFUN_DLD (__midicallback__, args, nargout,
"-*- texinfo -*-\n\
@deftypefn {Package} {} __midicallback__ ()\n \
Internal function for midicallback\n \
\n \
@end deftypefn")
{

  octave_value_list retval;

  if (1 > args.length ())
  {
    error ("Expected at least a midi device");
    return retval;
  }

  if (args (0).type_id () != octave_midi::static_type_id ())
    {
      error ("Expected midi device");
      return retval;
    }

  octave_midi * dev = dynamic_cast<octave_midi*>(args (0).internal_rep ());

  std::string callback = dev->get_callback();
  if (args.length () > 2)
    {
      if (!args(1).is_string())
        {
          error ("Expected callback as string");
          return retval;
        }
      // set
      dev->set_callback(args(1).string_value(), args(2));
    }
  else if (args.length () != 1)
    {
      error ("Expected mididevice with optional (callback, data)");
      return retval;
    }

  retval(0) = callback;

  return retval;
}
